File: libs/my-test-domain/infrastructure/src/lib/adapters/event_bus_in_memory_adapter.py
from typing import Type, Callable, Dict, List
from my-test-domain.domain.ports import IEventBus

class EventBusInMemoryAdapter(IEventBus):
    def __init__(self) -> None:
        self._handlers: Dict[Type, List[Callable]] = {}

    def publish(self, event: object) -> None:
        event_type = type(event)
        if event_type in self._handlers:
            for handler in self._handlers[event_type]:
                handler(event)

    def subscribe(self, event_type: Type, handler: Callable) -> None:
        if event_type not in self._handlers:
            self._handlers[event_type] = []
        self._handlers[event_type].append(handler)

File: libs/my-test-domain/infrastructure/src/lib/adapters/event-bus.in-memory.adapter.ts
import { IEventBus } from '@my-test-domain/domain';

export class EventBusInMemoryAdapter implements IEventBus {
  private handlers: Map<string, any[]> = new Map();

  publish(event: any): void {
    const eventName = event.constructor.name;
    const eventHandlers = this.handlers.get(eventName);
    if (eventHandlers) {
      eventHandlers.forEach((handler) => handler(event));
    }
  }

  subscribe(event: any, handler: any): void {
    const eventName = event.name;
    if (!this.handlers.has(eventName)) {
      this.handlers.set(eventName, []);
    }
    this.handlers.get(eventName).push(handler);
  }
}
